hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/lib/x-platform/c-api/dirent.cc

```diff
@@ -0,0 +1,100 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <cerrno>
+#include <iostream>
+#include <iterator>
+#include <system_error>
+#include <variant>
+
+#include "x-platform/c-api/dirent.h"
+#include "x-platform/dirent.h"
+
+#if defined(WIN32) && defined(__cplusplus)
```

Comment:
There's no cleaner way to do this?

Reply:
The C++ compiler will mangle the method signature. Since these APIs are invoked in jnihelper.c (which is compiled by C compiler), the C linker won't be able to find the mangled version of these methods. Thus, we need to wrap the implementation with an `extern "C"` block to prevent the C++ compiler from mangling these APIs. It results in an `Undefined reference error` otherwise.

This is the standard approach whenever C code tries to invoke C++ code.

Reply:
I'm OK with the extern, what looks weird is to use if/endif to wrap it.
Isn't there a cleaner way where we have the core in a header file and then we have the extern in some other and we do include or not accordingly?
```
extern "C" {
#include <xxx.h>
}
```

Reply:
I've put the `extern "C"` in the `c-api/dirent.h`. I thought that I had to enclose the implementation as well. I just removed the `extern "C"` enclosure in the implementation and I was able to build fine on both Windows and Ubuntu. I ran the tests on both the platforms and they're working fine -

## Windows 10
![image](https://user-images.githubusercontent.com/10280768/172365616-57f1b051-e010-4f18-bc99-344cfb1496af.png)

## Ubuntu Focal
![image](https://user-images.githubusercontent.com/10280768/172365730-df247cb8-8e7b-44f3-b094-a22dffa76d53.png)

I've removed the `extern "C"` enclosure from the implementation now, since it's redundant.